src/engine/MultiColumnJoin.cpp

```diff
@@ -0,0 +1,471 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@netpun.uni-freiburg.de)
+
+#include "./MultiColumnJoin.h"
+
+using std::string;
+
+// _____________________________________________________________________________
+MultiColumnJoin::MultiColumnJoin(QueryExecutionContext* qec,
+                                 std::shared_ptr<QueryExecutionTree> t1,
+                                 std::shared_ptr<QueryExecutionTree> t2,
+                                 const vector<array<Id, 2>>& jcs)
+    : Operation(qec), _joinColumns(jcs), _multiplicitiesComputed(false) {
+  // Make sure subtrees are ordered so that identical queries can be identified.
+  AD_CHECK_GT(jcs.size(), 0);
+  if (t1->asString() < t2->asString()) {
+    _left = t1;
+    _right = t2;
+  } else {
+    _left = t2;
+    _right = t1;
+    for (unsigned int i = 0; i < _joinColumns.size(); i++) {
+      std::swap(_joinColumns[i][0], _joinColumns[i][1]);
```

Comment:
Add a comment that we need to swap the `_joinColumns` to match the the swapped trees. I think it's not immediately obvious